kuma/core/middleware.py

```diff
@@ -1,93 +1,181 @@
+            # Language code is a specific locale (fr vs fr-FR)
+            fixed_locale = language_from_path
+
+        if fixed_locale:
+            # Replace the 404 with a redirect to the fixed locale
+            fixed_url = "%s://%s%s" % (
+                request.scheme,
+                request.get_host(),
+                request.get_full_path().replace(literal_from_path,
+                                                fixed_locale,
+                                                1)
+            )
+            redirect_response = HttpResponseRedirect(fixed_url)
+            add_shared_cache_control(redirect_response)
+            return redirect_response
+        else:
+            # No language fixup found, return the 404
             return response
 
-        request.path_info = '/' + prefixer.shortened_path
-        request.LANGUAGE_CODE = prefixer.locale or settings.LANGUAGE_CODE
-        # prefixer.locale can be '', but we need a real locale code to activate
-        # otherwise the request uses the previously handled request's
-        # translations.
-        translation.activate(prefixer.locale or settings.LANGUAGE_CODE)
+
+class LocaleMiddleware(object):
+    """
+    This is a very simple middleware that parses a request
+    and decides what translation object to install in the current
+    thread context. This allows pages to be dynamically
+    translated to the language the user desires (if the language
+    is available, of course).
+
+    Based on Django 1.8.19's LocaleMiddleware from
+    django/middleware/locale.py, with changes:
+
+    * Assume that locale prefixes are in use
+    * Use Kuma language codes (en-US) instead of Django's (en-us)
+    * Use Kuma-prefered locales (zn-CN) instead of Django's (zn-Hans)
+    * Don't include "Vary: Accept-Language" header
+    * Add caching headers to locale redirects
+    """
+    response_redirect_class = HttpResponseRedirect
+
+    def process_request(self, request):
+        """Activate the language, based on the request."""
+        language = get_language_from_request(request)
+        translation.activate(language)
+        request.LANGUAGE_CODE = language
 
     def process_response(self, request, response):
-        """Unset the thread-local var we set during `process_request`."""
-        # This makes mistaken tests (that should use LocalizingClient but
-        # use Client instead) fail loudly and reliably. Otherwise, the set
-        # prefixer bleeds from one test to the next, making tests
-        # order-dependent and causing hard-to-track failures.
-        set_url_prefixer(None)
+        """Add Content-Language, convert some 404s to locale redirects."""
+        language = get_language()
+        language_from_path = get_language_from_path(request.path_info)
+        if response.status_code == 404 and not language_from_path:
+            language_path = '/%s%s' % (language, request.path_info)
+            path_valid = is_valid_path(request, language_path)
```

Comment:
This fails to redirect locale-free zone URL's (e.g., `http://localhost:8000/Firefox` gets a 404 instead of redirecting to the active locale set from `get_language_from_request` on the way in), which causes a number of the tests within `tests/headless/test_cdn.py` to fail. It seems that `is_valid_path` has to consider zones as well? 🤦‍♂️ I'm tempted to say we should take this opportunity to dump zones now, but that probably is a bridge too far.

Reply:
Which tests fail?

Reply:
I posted my actual results in the main thread below.

Reply:
To summarize the results and discussion below:

* Tests against /es/Firefox and /fr/Firefox failed because only the English page is in the sample database. This code returns a 404, where the previous code returned a 302. This is probably an improvement, but means those tests should be skipped or marked ``XFAIL`` in local development, until the sample database is updated.
* The test for ``?lang=de`` succeeded with a 302, even though the page is not in the sample database.  The quick 404 would be nice, but it isn't a regression, so I'm not considering a change for this PR.